File: content/tracks/dsa/problems/lc-0739-daily-temperatures/solution.go
package main

func dailyTemperatures(temperatures []int) []int {
	answer := make([]int, len(temperatures))
	// Indices of days still waiting for a warmer day; their temperatures never increase
	// from bottom to top.
	waiting := []int{}
	for day, temp := range temperatures {
		for len(waiting) > 0 && temperatures[waiting[len(waiting)-1]] < temp {
			earlier := waiting[len(waiting)-1]
			waiting = waiting[:len(waiting)-1]
			answer[earlier] = day - earlier
		}
		waiting = append(waiting, day)
	}
	return answer
}

File: content/tracks/dsa/problems/lc-0739-daily-temperatures/solution.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        answer = [0] * len(temperatures)
        # Indices of days still waiting for a warmer day; their temperatures never increase
        # from bottom to top.
        waiting: List[int] = []
        for day, temp in enumerate(temperatures):
            while waiting and temperatures[waiting[-1]] < temp:
                earlier = waiting.pop()
                answer[earlier] = day - earlier
            waiting.append(day)
        return answer
